### PR TITLE
[Function builders] Infer function builder through @_dynamicReplacement(for:)

### PR DESCRIPTION
**Pull request by DougGregor**
Fixes rdar://problem/63898120.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -5075,8 +5075,8 @@ ERROR(function_builder_infer_ambig, none,
 NOTE(function_builder_infer_add_return, none,
      "add an explicit 'return' statement to not use a function builder", ())
 NOTE(function_builder_infer_pick_specific, none,
-     "apply function builder %0 (inferred from protocol %1)",
-     (Type, DeclName))
+     "apply function builder %0 (inferred from %select{protocol|dynamic replacement of}1 %2)",
```
Comment (hborla) [inline include/swift/AST/DiagnosticsSema.def:5078]: When is it possible for there to be an ambiguity between a function builder type inferred from a protocol requirement and one inferred from dynamic replacement? Does this happen if the dynamic replacement satisfies a protocol requirement, and both the requirement and the dynamically replaced decl specify different function builders?

**Review by hborla — APPROVED** [review]
LGTM! One small question below (and perhaps a test should be added for that case):

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (DougGregor) [reply include/swift/AST/DiagnosticsSema.def:5078]: Yeah, good point. I meant to have a test for that. I've added it in the latest commit.

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Review by xedin — APPROVED** [review]

**Pull request merged**